singularity-eos/eos/eos_eospac.hpp

```diff
@@ -89,86 +88,640 @@ class EOSPAC : public EosBase<EOSPAC> {
+    PressureFromDensityInternalEnergy(R, sie_p_de, Pe, internal_scratch, num, lambdas);
+    TemperatureFromDensityInternalEnergy(rho_p_drho, E, Tr, internal_scratch, num,
+                                         lambdas);
+    TemperatureFromDensityInternalEnergy(R, sie_p_de, Te, internal_scratch, num, lambdas);
+
+    portableFor(
+        cname, 0, num, PORTABLE_LAMBDA(const int i) {
+          dpdrs[i] = (Pr[i] - presses[i]) / drho[i];
+          dpdes[i] = (Pe[i] - presses[i]) / de[i];
+          dtdrs[i] = (Tr[i] - temps[i]) / drho[i];
+          dtdes[i] =
+              (Te[i] - temps[i]) /
+              de[i]; // Would it be better to skip the calculation of Te and return 1/cv?
+        });
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  void PTofRE(Real &rho, Real &sie, Real *lambda, Real &press, Real &temp, Real &dpdr,
+              Real &dpde, Real &dtdr, Real &dtde) const {
+
+    press = PressureFromDensityInternalEnergy(rho, sie, lambda);
+    temp = TemperatureFromDensityInternalEnergy(rho, sie, lambda);
+    const Real drho = rho * 1.0e-6;
+    const Real de = sie * 1.0e-6;
+    const Real Pr = PressureFromDensityInternalEnergy(rho + drho, sie, lambda);
+    const Real Pe = PressureFromDensityInternalEnergy(rho, sie + de, lambda);
+    const Real Tr = TemperatureFromDensityInternalEnergy(rho + drho, sie, lambda);
+    const Real Te = TemperatureFromDensityInternalEnergy(rho, sie + de, lambda);
+    dpdr = (Pr - press) / drho;
+    dpde = (Pe - press) / de;
+    dtdr = (Tr - temp) / drho;
+    dtde = (Te - temp) /
+           de; // Would it be better to skip the calculation of Te and return 1/cv?
+    return;
+  }
+
+  static inline unsigned long scratch_size(std::string method, unsigned int nelements) {
+    auto nbuffers = scratch_nbuffers();
+    if (nbuffers.find(method) != nbuffers.end()) {
+      return sizeof(Real) * nbuffers[method] * nelements;
+    }
+    return 0;
+  }
+
+  static inline unsigned long max_scratch_size(unsigned int nelements) {
+    auto nbuffers = scratch_nbuffers();
+    auto e = std::max_element(
+        nbuffers.begin(), nbuffers.end(),
+        [](const auto &a, const auto &b) { return a.second < b.second; });
+    return scratch_size(e->first, nelements);
+  }
+
+  static inline std::map<std::string, unsigned int> &scratch_nbuffers() {
+    static std::map<std::string, unsigned int> nbuffers = {
+        {"TemperatureFromDensityInternalEnergy", 2},
+        {"PressureFromDensityTemperature", 2},
+        {"FillEos", 4},
+        {"InternalEnergyFromDensityTemperature", 2},
+        {"PressureFromDensityInternalEnergy", 3},
+        {"SpecificHeatFromDensityTemperature", 2},
+        {"SpecificHeatFromDensityInternalEnergy", 3},
+        {"BulkModulusFromDensityTemperature", 6},
+        {"BulkModulusFromDensityInternalEnergy", 6},
+        {"GruneisenParamFromDensityTemperature", 4},
+        {"GruneisenParamFromDensityInternalEnergy", 5},
+        {"PTofRE", 11}};
+    return nbuffers;
```

Comment:
Since the scratch methods are available in the variant, should the scratch size methods be available too? We can implement the "null" versions in the base class and just return 0?

Reply:
I remember discussing this before and I thought we decided that it was specific enough to EOSPAC that a host code would need to differentiate anyway. However, I'm less convinced now that that's the right way to go, especially if we use EOSPAC in the PTE solvers. I'd support including null versions in the variant.

Reply:
Done.